### PR TITLE
drivers: spi: Added qspi driver for ATSAME and ATSAMD parts

### PR DESCRIPTION
**Pull request by flaminggoat**
This driver allows the use of the QSPI peripheral in SPI mode. See page 192 of the family datasheet https://ww1.microchip.com/downloads/en/DeviceDoc/SAM_D5x_E5x_Family_Data_Sheet_DS60001507G.pdf

The driver currently only supports SPI mode using the hardware CS pin.
I attached the logic scope of running the spi_flash sample program, can be viewed with https://sigrok.org/wiki/PulseView .

```
JEDEC SPI-NOR SPI flash testing
==========================

Test 1: Flash erase
Flash erase succeeded!

Test 2: Flash write
Attempting to write 4 bytes
Data read matches data written. Good!!
```

[pulseview.zip](https://github.com/zephyrproject-rtos/zephyr/files/6365785/pulseview.zip)

Signed-off-by: Theo Hussey <theo@open-cosmos.com>

### CODE REVIEW

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Seems fine to me.

One change however: s/CONFIG_QSPI_SAM0/CONFIG_SPI_SAM0_QSPI 
(it's still a SPI driver in the end)

Change the commit titles adding prefix "drivers: spi: "

**Comment by nandojve** [comment]
Hi @flaminggoat ,

Just to inform, it is not allowed merge branches in Zephyr. I recommend you rebase your code on top of main. You can read [How to Contribute](https://docs.zephyrproject.org/latest/contribute/index.html) to get a better view.

**Comment by tbursztyka** [comment]
Instead of merging which creates unecessary commits, rebase your branch against the main branch. And don't make commit with changes on top of original commit, rebase your branch and fixup your first commit.

File: dts/bindings/spi/atmel,sam0-spi-quadspi.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+    clock-names:
+      required: true
+      description: |
+        This should be "QSPI_AHB_Clock", "QSPI_2X_AHB_Clock", "QSPI_APB_Clock".
+
+        For example
+           clock-names = "QSPI_AHB_Clock", "QSPI_2X_AHB_Clock", "QSPI_APB_Clock";
+
+
```
Comment (nandojve) [inline dts/bindings/spi/atmel,sam0-spi-quadspi.yaml:25]: Drop lines

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @flaminggoat ,

I don't have any board to run local tests. I pointed some notes and I think you should squash all in only one commit. My concern is that no boards enabling this driver and we would not check at CI. Suggestion is add another commit that enable ATSAME54-XPRO QSPI.

**Comment by nandojve** [comment]
Hi @flaminggoat ,

Please, rebase your code and move to pinctrl API. See #43731 as reference.

**Comment by nandojve** [comment]
Hi @flaminggoat ,

Are you interested on this PR? I removed STALE so you can continue if you want.
Let me know if you still interested or not.

**Pull request closed without merging**